scipy/sparse/_index.py

```diff
@@ -0,0 +1,315 @@
+        row, col = self._validate_indices(key)
+
+        if isinstance(row, INT_TYPES) and isinstance(col, INT_TYPES):
+            x = np.asarray(x, dtype=self.dtype)
+            if x.size != 1:
+                raise ValueError('Trying to assign a sequence to an item')
+            self._set_intXint(row, col, x.flat[0])
+            return
+
+        if isinstance(row, slice):
+            row = np.arange(*row.indices(self.shape[0]))[:, None]
+        else:
+            row = np.atleast_1d(row)
+
+        if isinstance(col, slice):
+            col = np.arange(*col.indices(self.shape[1]))[None, :]
+            if row.ndim == 1:
+                row = row[:, None]
+        else:
+            col = np.atleast_1d(col)
+
+        i, j = np.broadcast_arrays(row, col)
+        if i.shape != j.shape:
+            raise IndexError('number of row and column indices differ')
+
+        from .base import isspmatrix
+        if isspmatrix(x):
+            if i.ndim == 1:
+                # Inner indexing, so treat them like row vectors.
+                i = i[None]
+                j = j[None]
+            broadcast_row = x.shape[0] == 1 and i.shape[0] != 1
+            broadcast_col = x.shape[1] == 1 and i.shape[1] != 1
+            if not ((broadcast_row or x.shape[0] == i.shape[0]) and
+                    (broadcast_col or x.shape[1] == i.shape[1])):
+                raise ValueError('shape mismatch in assignment')
+            if x.size == 0:
+                return
+            x = x.tocoo(copy=True)
+            x.sum_duplicates()
+            self._set_arrayXarray_sparse(i, j, x)
+        else:
+            # Make x and i into the same shape
+            x = np.asarray(x, dtype=self.dtype)
+            x, _ = np.broadcast_arrays(x, i)
+            if x.shape != i.shape:
+                raise ValueError("shape mismatch in assignment")
+            if x.size == 0:
+                return
+            self._set_arrayXarray(i, j, x)
+
+    def _validate_indices(self, key):
+        M, N = self.shape
+        row, col = _unpack_index(key)
+
+        if isintlike(row):
+            row = int(row)
+            if row < -M or row >= M:
+                raise IndexError('row index (%d) out of range' % row)
+            if row < 0:
+                row += M
+        elif not isinstance(row, slice):
+            row = self._asindices(row, M)
+
+        if isintlike(col):
+            col = int(col)
+            if col < -N or col >= N:
+                raise IndexError('column index (%d) out of range' % col)
+            if col < 0:
+                col += N
+        elif not isinstance(col, slice):
+            col = self._asindices(col, N)
+
+        return row, col
+
+    def _asindices(self, idx, length):
+        """Convert `idx` to a valid index for an axis with a given length.
+
+        Subclasses that need special validation can override this method.
+        """
+        try:
+            x = np.asarray(idx)
+        except:
```

Comment:
This bare `except` is getting flagged in testing

Reply:
My approach here and below was to replace the bare `except` with exceptions that I could find are somewhat-commonly raised with `np.asarray()` based on bug reports and so on. `MemoryError` feels like a bit of a stretch, but my assumption was that there was a desire to capture a broad set of exceptions here.

In making this change I also noted that `_asindices()` is only called twice in pure Python code in our code base here, and both times `idx` is coerced to an `int` beforehand anyway.

Presumably there are trickier scenarios where the indexing is used in the broader ecosystem, etc. If I'm missing a specific exception here I guess let me know--we should probably test for it too if that's important -- seems ok locally.